File: game/src/sdk/progression.ts
export const EXP_WIN = 50;
export const EXP_DRAW = 25;
export const EXP_LOSS = 15;
export const EXP_MOVE_CAP = 40;
export const MAX_LEVEL = 99;

export function expBaseForLevel(level: number): number {
  if (level <= 1) return 0;
  const capped = Math.min(level, MAX_LEVEL);
  return 50 * capped * (capped - 1);
}

export function expToNext(level: number): number {
  if (level < 1 || level >= MAX_LEVEL) return 0;
  return 100 * level;
}

export function levelFromExp(exp: number): number {
  if (!Number.isFinite(exp) || exp < 0) return 1;
  let level = 1;
  while (level < MAX_LEVEL && exp >= expBaseForLevel(level + 1)) {
    level += 1;
  }
  return level;
}

export interface LevelProgress {
  level: number;
  current: number;
  required: number;
  ratio: number;
}

export function levelProgress(exp: number): LevelProgress {
  const safeExp = Number.isFinite(exp) && exp > 0 ? exp : 0;
  const level = levelFromExp(safeExp);
  const required = expToNext(level);
  if (required <= 0) {
    return { level, current: 0, required: 0, ratio: 1 };
  }
  const current = safeExp - expBaseForLevel(level);
  return { level, current, required, ratio: Math.min(1, current / required) };
}

File: server/internal/game/progression/progression.go
package progression

const (
	ExpWin     = 50
	ExpDraw    = 25
	ExpLoss    = 15
	ExpMoveCap = 40
	MaxLevel   = 99
	ReasonVoid = "void"
)

func ExpBaseForLevel(level int) int64 {
	if level <= 1 {
		return 0
	}
	if level > MaxLevel {
		level = MaxLevel
	}
	return int64(50 * level * (level - 1))
}

func ExpToNext(level int) int64 {
	if level < 1 || level >= MaxLevel {
		return 0
	}
	return int64(100 * level)
}

func LevelFromExp(exp int64) int {
	if exp < 0 {
		return 1
	}
	level := 1
	for level < MaxLevel && exp >= ExpBaseForLevel(level+1) {
		level++
	}
	return level
}

func MoveBonus(moveCount int) int {
	if moveCount < 0 {
		return 0
	}
	if moveCount > ExpMoveCap {
		moveCount = ExpMoveCap
	}
	return moveCount / 2
}

func MatchExp(winnerID, reason, playerID string, moveCount int) int {
	if reason == ReasonVoid {
		return 0
	}
	base := ExpLoss
	if winnerID == "" {
		base = ExpDraw
	} else if playerID == winnerID {
		base = ExpWin
	}
	return base + MoveBonus(moveCount)
}
